lua/packer/luarocks.lua

```diff
@@ -0,0 +1,305 @@
+
+    local opts = {capture_output = callbacks}
+    local r = await(jobs.run(hererocks_cmd, opts)):map_err(
+                function(err)
+        return {msg = 'Error installing hererocks', data = err, output = output}
+      end)
+
+    local luarocks_cmd = 'python ' .. hererocks_file .. ' --verbose -j ' .. lua_version.jit
+                           .. ' -r latest ' .. hererocks_install_dir
+    r = r:and_then(await, jobs.run(luarocks_cmd, opts)):map_err(
+          function(err) return {msg = 'Error installing luarocks', data = err, output = output} end)
+    return r
+  end)
+end
+
+local function package_patterns(dir) return fmt('%s?.lua;%s&/init.lua', dir, dir) end
+local package_paths = (function()
+  local install_path = util.join_paths(hererocks_install_dir, 'lib', 'luarocks',
+                                       fmt('rocks-%s', lua_version.lua))
+  local share_path = util.join_paths(hererocks_install_dir, 'share', 'lua', lua_version.lua)
+  return package_patterns(share_path) .. ';' .. package_patterns(install_path)
+end)()
+
+local nvim_paths_are_setup = false
+local function setup_nvim_paths()
+  if not hererocks_is_setup() then
+    log.warning('Tried to setup Neovim Lua paths before hererocks was setup!')
+    return
+  end
+
+  if nvim_paths_are_setup then
+    log.warning('Tried to setup Neovim Lua paths redundantly!')
+    return
+  end
+
+  if not string.find(package.path, package_paths, 1, true) then
+    package.path = package.path .. ';' .. package_paths
+  end
+
+  local install_cpath = util.join_paths(hererocks_install_dir, 'lib', 'lua', lua_version.lua)
+  local install_cpath_pattern = fmt('%s?.so', install_cpath)
+  if not string.find(package.cpath, install_cpath_pattern, 1, true) then
+    package.cpath = package.cpath .. ';' .. install_cpath_pattern
+  end
+
+  nvim_paths_are_setup = true
+end
+
+local function generate_path_setup_code()
+  local package_path_str = vim.inspect(package_paths)
+  local install_cpath = util.join_paths(hererocks_install_dir, 'lib', 'lua', lua_version.lua)
+  local install_cpath_pattern = fmt('"%s?.so"', install_cpath)
+  return [[
+  local package_path_str = ]] .. package_path_str .. [[
+
+  local install_cpath_pattern = ]] .. install_cpath_pattern .. [[
+
+  if not string.find(package.path, package_path_str, 1, true) then
+    package.path = package.path .. ';' .. package_path_str
+  end
+
+  if not string.find(package.cpath, install_cpath_pattern, 1, true) then
+    package.cpath = package.cpath .. ';' .. install_cpath_pattern
+  end
+]]
+end
+
+local function activate_hererocks_cmd(install_path)
+  local activate_file = 'activate'
+  local user_shell = os.getenv('SHELL')
+  local shell = user_shell:gmatch('([^/]*)$')()
+  if shell == 'fish' then
+    activate_file = 'activate.fish'
+  elseif shell == 'csh' then
+    activate_file = 'activate.csh'
+  end
+
+  return fmt('source %s', util.join_paths(install_path, 'bin', activate_file))
+end
+
+local function run_luarocks(args, disp)
```

Comment:
I had dreams of importing luarocks here and running `require('luarocks.cmd').install(...)` or similar, but never panned out. Probably works better for you to use it via shell anyway tho so that it's async

Reply:
Yeah, so did I. Unfortunately, the LuaRocks folks say there's no official way to do that, and I haven't found an unofficial one (though I haven't exactly looked hard; like you say, being able to run it async is critical)